whelktool/scripts/cleanups/2022/09/lxl-3919-unlink-agents.groovy

```diff
@@ -0,0 +1,113 @@
+// Remove any links (contributions) to the listed records.
+
+String where = """
+        collection = 'bib' AND deleted = false AND
+        id in(
+        select id from lddb__dependencies where dependsonid in (
+        '53hlt8kp5swj700',
+        'zcr7x820wh3nsszg',
+        'jgvz5wj22slrnfq',
+        'hftx5b9109jr70t',
+        'rp3584g92s7sftz',
+        'r42bsxm6p5kzxfmb',
+        'sq47ch5b5mzxpwv',
+        'sq47ff6b3hctcww',
+        '42gksr6n3064169',
+        'mkz2bc150ws5vp0',
+        'c9psxdww3m4g1wm',
+        'zw9djxwh1qb17kr',
+        'b8nrzclv3b40c5k',
+        'dbqtzzmx2wzmhq4',
+        '97mq02bt3x8kp59',
+        '31fjps8m4cxxh4t',
+        'vs69hv4d3x2frg1',
+        '0xbd8xmj5mx9p79',
+        'khwzsvc335hbf0g',
+        'ljx194t4087wblf',
+        'ljx14nq40d15b9d',
+        '31fjng6m195sp2f',
+        '97mqtj9t3j06v6c',
+        'bqlt0l9b8ng2zjwz',
+        'v61b0qjlsz7kx55j',
+        'bmcrhjsj8zv7p1h4',
+        'q22jtchnnrx9s3pg',
+        '53hkf25p015s2zs',
+        '42gkng5n0zlwxp5',
+        '31fjn1gm5853bpr',
+        'ljx01184543708h',
+        '64jlgctq5gsj2ls',
+        '53hlvs7p3shfcqx',
+        '5k0ckm4r3t6mdq96',
+        'z8cng026wjksnklw',
+        'rp3551b939f7xhk',
+        'khw04l535sq2k52'
+        ) )
+"""
+
+selectBySqlWhere(where) { data ->
+    //String initial = data.doc.getDataAsString()
+    boolean changed = data.graph[1].instanceOf?.contribution?.removeAll { contribution ->
+        def agent = contribution.agent
+        if (agent != null && asList(agent["@id"])[0] != null) {
+            return isABadLink(asList(agent["@id"])[0])
+        }
+    }
+
+    if (data.graph[1].instanceOf?.contribution != null && data.graph[1].instanceOf?.contribution.isEmpty())
+        data.graph[1].instanceOf.remove("contribution")
+
+    if (changed) {
+        //System.err.println("result:\n" +initial + "\nchanged into:\n"+ data.doc.getDataAsString() + "\n\n")
+        data.scheduleSave()
+    }
+}
+
+boolean isABadLink(String candidate) {
+    String s = candidate.substring(0, candidate.length()-3) // Trim off the #it
+    System.err.println(candidate + " -> " + s)
+    return s.endsWith("53hlt8kp5swj700") ||
```

Comment:
Nit-pick: safer (and slightly faster) to put these in a `Set` and use that here and to build the select string.

Reply:
To use a Set I'd instead have to shave of the https://libris{-qa,-stg}/ (which varies in length), which could go wrong. I think I'll keep this as is!

Reply:
Or use `candidate.substring(candidate.lastIndexOf('/') + 1, candidate.lastIndexOf('#'))`?

Reply:
`candidate.split('[#/]')[-2]` should work too!

Reply:
Yes, that is certainly much more succinct (albeit somewhat less performant, as it would parse a regexp (unless cached under the hood), splitting on all `/`:s and building a new list of those substrings; though I bet any performance loss is dwarfed by the I/O going on here).